primitives/file-key-verifier/src/lib.rs

```diff
@@ -0,0 +1,160 @@
+
+use codec::{Decode, Encode};
+use frame_support::sp_runtime::DispatchError;
+use num_bigint::BigUint;
+use scale_info::TypeInfo;
+use shp_traits::AsCompact;
+use sp_core::Hasher;
+use sp_std::{collections::btree_set::BTreeSet, vec::Vec};
+use sp_trie::{CompactProof, TrieDBBuilder, TrieLayout};
+use storage_hub_traits::CommitmentVerifier;
+use trie_db::Trie;
+
+#[cfg(test)]
+mod tests;
+
+/// A struct that implements the `CommitmentVerifier` trait, where the commitment
+/// is a Merkle Patricia Trie root hash and the response to a challenge is given
+/// by taking the modulo of the challenged hash with the number of chunks in the file,
+/// and interpreting the result as a chunk index.
+pub struct FileKeyVerifier<
+    T: TrieLayout,
+    const H_LENGTH: usize,
+    const CHUNK_SIZE: u64,
+    const SIZE_TO_CHALLENGES: u64,
+> where
+    <T::Hash as sp_core::Hasher>::Out: for<'a> TryFrom<&'a [u8; H_LENGTH]>,
+{
+    pub _phantom: core::marker::PhantomData<T>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, TypeInfo, Encode, Decode)]
+pub struct FileKeyProof {
+    pub owner: Vec<u8>,
+    pub location: Vec<u8>,
+    pub size: u64,
+    pub fingerprint: [u8; 32],
+    pub proof: CompactProof,
+}
+
+/// Implement the `CommitmentVerifier` trait for the `FileKeyVerifier` struct.
+impl<
+        T: TrieLayout,
+        const H_LENGTH: usize,
+        const CHUNK_SIZE: u64,
+        const SIZE_TO_CHALLENGES: u64,
+    > CommitmentVerifier for FileKeyVerifier<T, H_LENGTH, CHUNK_SIZE, SIZE_TO_CHALLENGES>
+where
+    <T::Hash as sp_core::Hasher>::Out: for<'a> TryFrom<&'a [u8; H_LENGTH]>,
+{
+    type Proof = FileKeyProof;
+    type Commitment = <T::Hash as sp_core::Hasher>::Out;
+    type Challenge = <T::Hash as sp_core::Hasher>::Out;
+
+    /// Verifies a proof against a root (i.e. commitment) and a set of challenges.
+    ///
+    /// Iterates over the challenges, computes the modulo of the challenged hashes with the number of chunks in the file,
+    /// and checks if the resulting leaf is in the proof.
+    fn verify_proof(
+        expected_file_key: &Self::Commitment,
+        challenges: &[Self::Challenge],
+        proof: &Self::Proof,
+    ) -> Result<Vec<Self::Challenge>, DispatchError> {
+        // Check that `challenges` is not empty.
+        if challenges.is_empty() {
+            return Err("No challenges provided.".into());
+        }
+
+        // Construct file key from the fields in the proof.
+        let file_key = T::Hash::hash(
+            &[
+                &proof.owner.encode(),
+                &proof.location.encode(),
+                &AsCompact(proof.size).encode(),
+                &proof.fingerprint.encode(),
+            ]
+            .into_iter()
+            .flatten()
+            .cloned()
+            .collect::<Vec<u8>>(),
+        );
+
+        // Check that the number of challenges is proportional to the size of the file.
+        let mut chunks_to_check = proof.size / SIZE_TO_CHALLENGES;
```

Comment:
This looks like it's going to be reused too, i.e. `FileMetadata::chunks_to_challenge(&self);`.